### PR TITLE
Implement Encryption

### PR DESCRIPTION
**Pull request by solokyo**

### CODE REVIEW

**Comment by jesperpedersen** [comment]
You need to open an issue with a short description of the work in this pull request, and possible follow-ups.

You can create a checklist in the issue of tasks that needs to be worked on

**Comment by solokyo** [comment]
Sure, I created an issues page #91

**Comment by jesperpedersen** [comment]
Ok, so now you can squash and reference #91 from your commit message.

See CONTRIBUTING.md for the commands

**Comment by jesperpedersen** [comment]
Remember to run `uncrustify` before the next `push -f`

**Comment by solokyo** [comment]
Please take another look.

File: src/libpgmoneta/wf_encryption.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+struct workflow*
+pgmoneta_workflow_encryption(bool encrypt)
+{
+   struct workflow* wf = NULL;
+
+   wf = (struct workflow*)malloc(sizeof(struct workflow));
+
+   wf->setup = &encryption_setup;
+
+   if (encrypt == true)
```
Comment (jesperpedersen) [inline src/libpgmoneta/wf_encryption.c:55]: Just `if (encrypt)`

File: src/libpgmoneta/wf_encryption.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+   config = (struct configuration*)shmem;
+
+   if (i_nodes != NULL)
+   {
+      prefix = pgmoneta_get_node_string(i_nodes, "prefix");
+
+      if (!strcmp(prefix, "Restore"))
+      {
+         to = pgmoneta_get_node_string(*o_nodes, "to");
+         d = malloc(strlen(to) + 1);
```
Comment (jesperpedersen) [inline src/libpgmoneta/wf_encryption.c:103]: You can use `pgmoneta_append()` here

File: src/libpgmoneta/workflow.c [hunk 3]
```diff
@@ -206,6 +218,12 @@ wf_archive(void)
       current = current->next;
    }
 
+   if (config->encryption != ENCRYPTION_NONE)
```
Comment (jesperpedersen) [inline src/libpgmoneta/workflow.c:221]: I don't get a `.zstd.aes` archive

File: src/libpgmoneta/aes.c [hunk 4]
```diff
@@ -27,57 +27,281 @@
...
+            to = NULL;
+
+            to = pgmoneta_append(to, d);
+            to = pgmoneta_append(to, "/");
+            to = pgmoneta_append(to, entry->d_name);
+            to = pgmoneta_append(to, ".aes");
+
+            if (pgmoneta_exists(from))
+            {
+               pgmoneta_log_debug("encrypting from %s to %s", from, to);
```
Comment (jesperpedersen) [inline src/libpgmoneta/aes.c:94]: Remove this `debug`

File: src/libpgmoneta/aes.c [hunk 5]
```diff
@@ -27,57 +27,281 @@
...
+   closedir(dir);
+   return 0;
+}
 
-// [public]
 int
-pgmoneta_encrypt(char* plaintext, char* password, char** ciphertext, int* ciphertext_length)
+pgmoneta_encrypt_file(char* from, char* to)
+{
+   pgmoneta_log_debug("decrypting from %s to %s", from, to);
```
Comment (jesperpedersen) [inline src/libpgmoneta/aes.c:185]: Remove this `debug`

File: src/libpgmoneta/aes.c [hunk 6]
```diff
@@ -27,57 +27,281 @@
...
+   }
+
+   closedir(dir);
+   return 0;
+}
+
+int
+pgmoneta_encrypt_wal(char* d)
+{
+   pgmoneta_log_info("encrypt_wal:d=%s", d);
```
Comment (jesperpedersen) [inline src/libpgmoneta/aes.c:112]: Remove this `info`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
The archive feature doesn't encrypt here, using

```
pgmoneta-cli archive primary newest current /tmp
```

**Comment by jesperpedersen** [comment]
As noted on the issue the documentation is missing.

It is important that `doc/ENCRYPTION.md` describes which algorithms are supported, and why. Plus give the reader an idea of the overhead between `none` and the supported algorithms.

**Comment by solokyo** [comment]
I will let you know once I finished the modification.

**Comment by solokyo** [comment]
I pushed doc into a new commit, after we discuss and fix it, I will squash it with the previous commit.

File: doc/ENCRYPTION.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
+## Overview
+AES Cipher block chaining (CBC) mode and AES Counter (CTR) mode are supported in pgmoneta. The default setup is no encryption. 
+
+CBC is the most commonly used and considered save mode. Its main drawbacks are that encryption is sequential (decryption can be parallelized).
+
+Along with CBC, CTR mode is one of two block cipher modes recommended by Niels Ferguson and Bruce Schneier. Both encryption and decryption are parallelizable.
+
+Longer the key length, safer the encryption. However, with 20% (192 bit) and 40 (256 bit) extra workload compare to 128 bit.
```
Comment (jesperpedersen) [inline doc/ENCRYPTION.md:8]: 40 -> 40%

File: doc/ENCRYPTION.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
```
Comment (jesperpedersen) [inline doc/ENCRYPTION.md:9]: I would list the available options here as well, and highlight that `aes` is `aes-256-cbc`

File: doc/man/pgmoneta.conf.5.rst [hunk 9]
```diff
@@ -45,6 +45,9 @@ compression
 compression_level
   The compression level. Default is 3
 
+encryption
+  The encryption mode. Default is none
```
Comment (jesperpedersen) [inline doc/man/pgmoneta.conf.5.rst:49]: Copy the description from CONFIGURATION.md without the formatting

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
You will need

* README.md: AES encryption support
* AUTHORS: Your name and email

**Comment by jesperpedersen** [comment]
Rebase and squash as well

**Comment by jesperpedersen** [comment]
Note, that archive would benefit from a `pgmoneta-cli decrypt <file>` command. This could be a separate pull request

**Comment by jesperpedersen** [comment]
Merged with a few modifications.

Thanks for your contribution !

**Pull request closed without merging**